Implement `fetch_values` for `HashWithIndifferentAccess`

joshpencheon opened the pull request:
`fetch_values` was added to `Hash` in Ruby 2.3.0: https://bugs.ruby-lang.org/issues/10017

This patch adds an implemention for instances of `HashWithIndifferentAccess`, in line with the existing definitions of `fetch` and `values_at`.

_Current behaviour, without patch:_
```ruby
hash = ActiveSupport::HashWithIndifferentAccess.new
hash[:a] = 'x'
hash[:b] = 'y'
hash.fetch_values('a', 'b') # => KeyError: key not found: "a"
```

_New behaviour, with patch:_
```ruby
hash = ActiveSupport::HashWithIndifferentAccess.new
hash[:a] = 'x'
hash[:b] = 'y'
hash.fetch_values('a', 'b') # => ["x", "y"]
hash.fetch_values('a', 'c') { |key| 'z' } # => ["x", "z"]
hash.fetch_values('a', 'c') # => KeyError: key not found: "c"
```

Thanks, Josh

joshpencheon committed:
Implement `fetch_values` for HashWithIndifferentAccess

`fetch_values` was added to Hash in Ruby 2.3.0:
  https://bugs.ruby-lang.org/issues/10017

This patch adds an implemention for instances of HWAI, in line
with the existing definitions of `fetch` and `values_at`.

matthewd commented on activesupport/lib/active_support/hash_with_indifferent_access.rb:
```diff
@@ -195,6 +195,19 @@ def values_at(*indices)
+    #   hash.fetch_values('a', 'c') # => KeyError: key not found: "c"
+    def fetch_values(*indices, &block)
+      indices.collect { |key| fetch(key, &block) }
+    end
```
Let's throw an `if Hash.method_defined? :fetch_values` onto the `end` here, so we don't get ahead of Hash.

(And a skip in the test to match)

joshpencheon commented:
Thanks @matthewd, I hadn't considered that. I've made the changes you suggested; is there any formal way of documenting that these clauses can be removed once Ruby 2.3 is the minimum requirement?

I've rebased now development has moved to 5.2, and squashed my commits. Is there anything this PR needs?

matthewd commented:
Thanks!

> is there any formal way of documenting that these clauses can be removed once Ruby 2.3 is the minimum requirement

Nope. I don't think we have enough of them to warrant a formal process... we'll just look for version-relevant conditionals (like `method_defined?` calls) when the time comes.

sandstrom commented:
I was just about to send a PR when I saw this! 😄  Great addition @joshpencheon!
